### PR TITLE
[#4169] feat(core): Supports the core logic of the ownership

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?
 Supports the core logic of the ownership.
 This pull request includes two parts:
 1. Add a new type relational interface for entity store
 2. Add the ownership manager and related database ops

### Why are the changes needed?

Fix: #4169

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Add the uts.

### CODE REVIEW

**Comment by jerqi** [comment]
@jerryshao @yuqi1129 @xunliu Could you take a look when you are free?

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    } catch (IOException ioe) {
+      LOG.info("Fail to set the owner {} of entity {}", ownerName, identifier, ioe);
+      throw new RuntimeException(ioe);
+    }
+  }
+
+  public Owner getOwner(NameIdentifier identifier, Entity.EntityType type) {
+    try {
+      OwnerImpl owner = new OwnerImpl();
+      List entities =
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java:86]: Better change to `List<Entity>`?

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import org.apache.gravitino.exceptions.NoSuchEntityException;
+import org.apache.gravitino.exceptions.NotFoundException;
+import org.apache.gravitino.exceptions.OwnerNotFoundException;
+import org.apache.gravitino.meta.GroupEntity;
+import org.apache.gravitino.meta.UserEntity;
+import org.apache.gravitino.relation.Relation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OwnershipManager {
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java:35]: Please add comments for the class OwnershipManager.

File: core/src/main/java/org/apache/gravitino/relation/Relation.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.gravitino.relation;
+
+/** Relation is an abstraction which connects two entities. */
+public interface Relation {
+
+  enum Type {
+    OWNER_REL
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/relation/Relation.java:25]: Which means `REL`, Please add a comment.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/OwnerMetaService.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+import org.apache.gravitino.Entity;
+import org.apache.gravitino.NameIdentifier;
+import org.apache.gravitino.exceptions.NoSuchEntityException;
+import org.apache.gravitino.storage.relational.mapper.OwnerMetaMapper;
+import org.apache.gravitino.storage.relational.po.OwnerRelPO;
+import org.apache.gravitino.storage.relational.utils.POConverters;
+import org.apache.gravitino.storage.relational.utils.SessionUtils;
+
+public class OwnerMetaService {
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/OwnerMetaService.java:29]: Please add a comment for OwnerMetaService.

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java (hunk 1 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java:87]: It will report error.

File: api/src/main/java/org/apache/gravitino/authorization/Owner.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.gravitino.authorization;
+
+/**
+ * Every metadata object has owner. Every role, user or group has owner, too. The owner can have all
```
Comment (yuqi1129) [inline api/src/main/java/org/apache/gravitino/authorization/Owner.java:22]: Has an owner?

File: api/src/main/java/org/apache/gravitino/authorization/Owner.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.gravitino.authorization;
+
+/**
+ * Every metadata object has owner. Every role, user or group has owner,too. The owner can have all
```
[the remark this answers is not in the document]
Comment (jerqi) [reply api/src/main/java/org/apache/gravitino/authorization/Owner.java]: Done.

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java (hunk 2 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java:36]: Done.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/OwnerMetaService.java (hunk 4 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/OwnerMetaService.java:32]: Added.

File: core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java [hunk 7]
```diff
@@ -131,58 +132,127 @@ public static <T> void prepareAuthorizationHooks(T manager, DispatcherHooks hook
       hooks.addPostHook(
           "createMetalake",
           (args, metalake) -> {
-            // TODO: Add the logic of setting the owner
+            NameIdentifier identifier = (NameIdentifier) ((Object[]) args)[0];
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java:135]: It's a little hacky to get instance of NameIdentifier here, how can I know the type of the first argument is `NameIdentifier`?

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  public Owner getOwner(NameIdentifier identifier, Entity.EntityType type) {
+    try {
+      OwnerImpl owner = new OwnerImpl();
+      List entities =
+          store
+              .relationOperations()
+              .listEntitiesByRelation(Relation.Type.OWNER_REL, identifier, type);
+      if (entities.isEmpty()) {
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java:91]: I'm afriad that existing entities may has no ower and then throw an exception here.

File: api/src/main/java/org/apache/gravitino/authorization/Owner.java (hunk 5 above)
Comment (jerqi) [reply api/src/main/java/org/apache/gravitino/authorization/Owner.java:22]: Fixed.

File: core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java (hunk 7 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java:135]: You can know it through the method `createMetalake`.

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java (hunk 8 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java:91]: Yes, it will throw an exception. We should return the exception to the client, too.

File: core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java (hunk 7 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java:135]: So if we change the method `createMetalake` and the type of first argument is NOT `NameIdentifier`,  we could not detect the error through compilation stage.

Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java:135]: But we can add some e2e test to detect it.

File: core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java [hunk 9]
```diff
@@ -131,58 +132,127 @@ public static <T> void prepareAuthorizationHooks(T manager, DispatcherHooks hook
       hooks.addPostHook(
           "createMetalake",
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java:133]: hi @jerqi  I think using function name string comparing is too error-prone.

We can create a `SetOwner` annotation to supports, for example:
```
public class MetalakeManager implements MetalakeDispatcher {
  @Override
  @SetOwner
  public BaseMetalake createMetalake(
      NameIdentifier ident, String comment, Map<String, String> properties)
      throws MetalakeAlreadyExistsException {
    long uid = idGenerator.nextId();
   ......
```

Use `SetOwner` annotation to comparing
```
if (method.isAnnotationPresent(SetOwner.class)) { 
NameIdentifier identifier = (NameIdentifier) ((Object[]) args)[0];
                            GravitinoEnv.getInstance()
                                    .ownershipManager()
                                    .setOwner(
                                            identifier,
                                            Entity.EntityType.METALAKE,
                                            PrincipalUtils.getCurrentUserName(),
                                            Owner.Type.USER);
}
```

File: core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java [hunk 10]
```diff
@@ -131,58 +132,127 @@ public static <T> void prepareAuthorizationHooks(T manager, DispatcherHooks hook
...
           (args, metalake) -> {
-            // TODO: Add the logic of setting the owner
+            NameIdentifier identifier = (NameIdentifier) ((Object[]) args)[0];
+            GravitinoEnv.getInstance()
+                .ownershipManager()
+                .setOwner(
+                    identifier,
+                    Entity.EntityType.METALAKE,
+                    PrincipalUtils.getCurrentUserName(),
+                    Owner.Type.USER);
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java:142]: We should not put important logic into the util class.

File: core/src/main/java/org/apache/gravitino/relation/SupportsRelationOperations.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.List;
+import org.apache.gravitino.Entity;
+import org.apache.gravitino.HasIdentifier;
+import org.apache.gravitino.NameIdentifier;
+
+/**
+ * This is an extended interface. This is mainly used for strengthen the ability of querying
+ * relational data.
+ */
+public interface SupportsRelationOperations {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/relation/SupportsRelationOperations.java:31]: We can split this into a different PR.

File: core/src/test/resources/h2/schema-h2.sql [hunk 12]
```diff
@@ -246,3 +246,20 @@ CREATE TABLE IF NOT EXISTS `tag_relation_meta` (
...
     KEY `idx_mid` (`metadata_object_id`)
     ) ENGINE=InnoDB;
+
+CREATE TABLE IF NOT EXISTS `owner_meta` (
+    `id` BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT COMMENT 'auto increment id',
+    `metalake_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'metalake id',
+    `owner_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'owner id',
+    `owner_type` VARCHAR(64) NOT NULL COMMENT 'owner type',
+    `entity_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'entity id',
+    `entity_type` VARCHAR(64) NOT NULL COMMENT 'entity type',
```
Comment (jerryshao) [inline core/src/test/resources/h2/schema-h2.sql:256]: owner is binding to a securable object or metadata object, you'd better use such name, not "entity"

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+public class OwnershipManager {
+  private static final Logger LOG = LoggerFactory.getLogger(OwnershipManager.class);
+  private final EntityStore store;
+
+  public OwnershipManager(EntityStore store) {
+    this.store = store;
+  }
+
+  public void setOwner(
+      NameIdentifier identifier, Entity.EntityType type, String ownerName, Owner.Type ownerType) {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java:45]: Owner is binding to the metadata/securable object, from interface level you should have such concept.

File: core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java (hunk 9 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java]: I will implement different dispatchers to refactor this code in the next pull request.

**Comment by jerqi** [comment]
@jerryshao @xunliu 
I have polished the code according to your suggestion.

File: core/src/test/resources/h2/schema-h2.sql (hunk 12 above)
Comment (jerqi) [reply core/src/test/resources/h2/schema-h2.sql]: I have changed this to metadata object.

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java (hunk 13 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java]: I changed to meta data object.

File: core/src/main/java/org/apache/gravitino/relation/SupportsRelationOperations.java (hunk 11 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/relation/SupportsRelationOperations.java:31]: If we split it into an another pull request, we can't get the complete mind.

File: core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java (hunk 10 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/AuthorizationUtils.java]: I will refactor the code using writing another dispatcher.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+          + " metadata_object_id as metadataObjectId,"
+          + " metadata_object_type as metadataObjectType,"
+          + " audit_info as auditInfo,"
+          + " current_version as currentVersion, last_version as lastVersion,"
+          + " deleted_at as deletedAt"
+          + " FROM "
+          + OWNER_TABLE_NAME
+          + " WHERE metadata_object_id = #{metadataObjectId}"
+          + " AND deleted_at = 0")
+  OwnerRelPO selectOwnerMetaByEntityIdAndType(@Param("metadataObjectId") Long metadataObjectId);
```
Comment (xloya) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java:53]: Maybe here need pass the `metadataType`.

File: core/src/test/resources/h2/schema-h2.sql [hunk 15]
```diff
@@ -246,3 +246,20 @@ CREATE TABLE IF NOT EXISTS `tag_relation_meta` (
...
+    `owner_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'owner id',
+    `owner_type` VARCHAR(64) NOT NULL COMMENT 'owner type',
+    `metadata_object_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'metadata object id',
+    `metadata_object_type` VARCHAR(64) NOT NULL COMMENT 'metadata object type',
+    `audit_info` MEDIUMTEXT NOT NULL COMMENT 'owner relation audit info',
+    `current_version` INT UNSIGNED NOT NULL DEFAULT 1 COMMENT 'owner relation current version',
+    `last_version` INT UNSIGNED NOT NULL DEFAULT 1 COMMENT 'owner relation last version',
+    `deleted_at` BIGINT(20) UNSIGNED NOT NULL DEFAULT 0 COMMENT 'owner relation deleted at',
+    PRIMARY KEY (`id`),
+    UNIQUE KEY `uk_ow_me_del` (`owner_id`, `metadata_object_id`, `deleted_at`),
```
Comment (xloya) [inline core/src/test/resources/h2/schema-h2.sql:262]: I think we need to add `metadata_type` here and below to ensure the uniqueness of metadata object. WDYT?

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+          + FilesetMetaMapper.META_TABLE_NAME
+          + " ft where ft.schema_id = #{schemaId} AND ft.deleted_at = 0 AND ot.deleted_at = 0 AND ft.fileset_id = ot.metadata_object_id"
+          + ")")
+  void sotDeleteOwnerRelBySchemaId(@Param("schemaId") Long schemaId);
+
+  @Delete(
+      "DELETE FROM "
+          + OWNER_TABLE_NAME
+          + " WHERE deleted_at > 0 AND deleted_at < #{legacyTimeline} LIMIT #{limit}")
+  Integer deleteOwnerMetasByLegacyTimeline(
```
Comment (xloya) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java:145]: This SQL is used in the `RelationalGarbageCollector`, maybe you need add some codes to call this SQL.

File: core/src/test/resources/h2/schema-h2.sql (hunk 15 above)
Comment (jerqi) [reply core/src/test/resources/h2/schema-h2.sql:262]: Our id is uuid. So it's not necessary to use the type.

File: core/src/main/java/org/apache/gravitino/storage/relational/utils/POConverters.java [hunk 17]
```diff
@@ -1055,8 +1055,8 @@ public static OwnerRelPO initializeOwnerRelPOsWithVersion(
           .withMetalakeId(metalakeId)
           .withOwnerId(ownerId)
           .withOwnerType(ownerType)
-          .withEntityId(entityId)
-          .withEntityType(entityType)
+          .withMetadataObjectId(entityId)
+          .withMetadataObjectType(entityType)
```
[the remark this answers is not in the document]
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/utils/POConverters.java:1059]: OK.

File: core/src/test/resources/h2/schema-h2.sql (hunk 15 above)
Comment (jerqi) [reply core/src/test/resources/h2/schema-h2.sql:262]: OK , I can align to other code.

File: core/src/main/java/org/apache/gravitino/relation/SupportsRelationOperations.java (hunk 11 above)
Comment (jerryshao) [reply core/src/main/java/org/apache/gravitino/relation/SupportsRelationOperations.java:31]: This `Relation` may be confused with `RelationalXXX`, maybe we should figure out a better name.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/UserMetaService.java [hunk 18]
```diff
@@ -151,10 +150,7 @@ public boolean deleteUser(NameIdentifier identifier) {
                 UserMetaMapper.class, mapper -> mapper.softDeleteUserMetaByUserId(userId)),
         () ->
             SessionUtils.doWithoutCommit(
-                UserRoleRelMapper.class, mapper -> mapper.softDeleteUserRoleRelByUserId(userId)),
-        () ->
-            SessionUtils.doWithoutCommit(
-                OwnerMetaMapper.class, mapper -> mapper.softDeleteOwnerRelByEntityId(userId)));
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/UserMetaService.java]: Why do we need to delete this?

File: core/src/main/java/org/apache/gravitino/storage/relational/service/GroupMetaService.java [hunk 19]
```diff
@@ -166,10 +165,7 @@ public boolean deleteGroup(NameIdentifier identifier) {
         () ->
             SessionUtils.doWithoutCommit(
                 GroupRoleRelMapper.class,
-                mapper -> mapper.softDeleteGroupRoleRelByGroupId(groupId)),
-        () ->
-            SessionUtils.doWithoutCommit(
-                OwnerMetaMapper.class, mapper -> mapper.softDeleteOwnerRelByEntityId(groupId)));
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/GroupMetaService.java]: Why do we need deleted this?

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java (hunk 2 above)
Comment (jerryshao) [reply core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java:37]: I would suggest to change to `OwnerManager` for unification.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/UserMetaService.java (hunk 18 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/UserMetaService.java]: User and group don't have ownership relation.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+          + " UNION "
+          + " SELECT tat.catalog_id FROM "
+          + TableMetaMapper.TABLE_NAME
+          + " tat where tat.catalog_id = #{catalogId} AND tat.deleted_at = 0 AND ot.deleted_at = 0 AND "
+          + "tat.table_id = ot.metadata_object_id AND ot.metadata_object_type = 'TABLE'"
+          + " UNION "
+          + " SELECT ft.catalog_id FROM "
+          + FilesetMetaMapper.META_TABLE_NAME
+          + " ft where ft.catalog_id = #{catalogId} AND ft.deleted_at = 0 AND ot.deleted_at = 0 AND"
+          + " ft.fileset_id = ot.metadata_object_id AND ot.metadata_object_type = 'FILESET'"
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java:123]: Change `where` to uppercase, also double check if all the reserved words are uppercase or not.

Comment (xunliu) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java:123]: @jerqi  Also have a `where` in the `org/apache/gravitino/storage/relational/mapper/SecurableObjectMapper.java`, Please change it to uppercase.

Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java:123]: Done.

File: core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java (hunk 2 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/OwnershipManager.java:44]: Done.

File: core/src/main/java/org/apache/gravitino/GravitinoEnv.java [hunk 21]
```diff
@@ -281,6 +283,10 @@ public TagManager tagManager() {
     return tagManager;
   }
 
+  public OwnerManager ownershipManager() {
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/GravitinoEnv.java:286]: Please change to `ownerManager()`

Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/GravitinoEnv.java]: Done.

**Review by xunliu — COMMENTED** [review]
LGTM

File: core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+          ioe);
+      throw new RuntimeException(ioe);
+    }
+  }
+
+  public Optional<Owner> getOwner(String metalake, MetadataObject metadataObject) {
+    try {
+      OwnerImpl owner = new OwnerImpl();
+      NameIdentifier ident = MetadataObjectUtil.toEntityIdent(metalake, metadataObject);
+      List entities =
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java:133]: Why do we use List without type, I think it has compile warning.

File: core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+        owner.type = Owner.Type.USER;
+      } else if (entities.get(0) instanceof GroupEntity) {
+        GroupEntity group = (GroupEntity) entities.get(0);
+        owner.name = group.name();
+        owner.type = Owner.Type.GROUP;
+      } else {
+        throw new IllegalArgumentException(
+            String.format(
+                "Doesn't support owner entity class %s", entities.get(0).getClass().getName()));
+      }
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java:166]: What is the meaning of `else` here, seems like this `else` will catch everything and throw an exception, how will the code reach to line 168?

File: core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java [hunk 24]
```diff
@@ -360,6 +363,39 @@ public List<TagEntity> associateTagsWithMetadataObject(
...
+      List<E> list = Lists.newArrayList();
+      OwnerMetaService.getInstance()
+          .getOwner(nameIdentifier, identType)
+          .ifPresent(e -> list.add((E) e));
+      return list;
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Doesn't support the relation type %s", relType));
+    }
+  }
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java:381]: Using switch case, not if else here and below.

File: core/src/main/java/org/apache/gravitino/storage/relational/po/OwnerRelPO.java [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+      validate();
+      return ownerRelPO;
+    }
+
+    private void validate() {
+      Preconditions.checkArgument(ownerRelPO.ownerId != null, "Owner id is required");
+      Preconditions.checkArgument(
+          StringUtils.isNotBlank(ownerRelPO.ownerType), "Owner type is required");
+      Preconditions.checkArgument(ownerRelPO.metadataObjectId != null, "Entity id is required");
+      Preconditions.checkArgument(ownerRelPO.metadataObjectType != null, "Entity type is required");
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/po/OwnerRelPO.java:107]: metadata object id and type, the string is not correct.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/OwnerMetaService.java [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+      return Optional.of(
+          UserMetaService.getInstance()
+              .getUserById(getMetalake(identifier), ownerRelPO.getOwnerId()));
+    } else if (ownerRelPO.getOwnerType().equals(Entity.EntityType.GROUP.name())) {
+      return Optional.of(
+          GroupMetaService.getInstance()
+              .getGroupById(getMetalake(identifier), ownerRelPO.getOwnerId()));
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Owner type doesn't support %s", ownerRelPO.getOwnerType()));
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/OwnerMetaService.java:66]: use switch case.

File: core/src/main/java/org/apache/gravitino/storage/relational/po/OwnerRelPO.java (hunk 25 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/po/OwnerRelPO.java:107]: I will correct and check all the changes again.

File: core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java (hunk 22 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java:133]: If I use entity here, it will throw an error.

File: core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java (hunk 23 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java:166]: If the entity is UserEntity or GroupEntity, the code won't reach the `else` code block.

File: core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java (hunk 24 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java:381]: OK, IDEA suggested I use `if else`. It's ok for me to use switch.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/OwnerMetaService.java (hunk 26 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/OwnerMetaService.java]: Changed this and Line 99.

File: core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java (hunk 24 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java:382]: Changed.

File: core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java (hunk 22 above)
Comment (xunliu) [reply core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java:133]: Is that ok?
```
List<? extends Entity> entities = 
```

Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java]: OK, It can compile.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+
+  @Update(
+      "UPDATE "
+          + OWNER_TABLE_NAME
+          + " SET deleted_at = (UNIX_TIMESTAMP() * 1000.0)"
+          + " + EXTRACT(MICROSECOND FROM CURRENT_TIMESTAMP(3)) / 1000"
+          + " WHERE metadata_object_id = #{metadataObjectId} AND metadata_object_type = #{metadataObjectType} AND deleted_at = 0")
+  void softDeleteOwnerRelByMetadataObjectIdAndType(
+      @Param("metadataObjectId") Long metadataObjectId,
+      @Param("metadataObjectType") String metadataObjectType);
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java:84]: Can we use `MetadataObject.type` type instead `String` for `metadataObjectType` param?
Beacuse if we can use `MetadataObject.type` type, the type parameters are very clear, 
and we can only call once `MetadataObject.type.name()` function, currently we need to call `.name()` function repeatedly.

Comment (qqqttt123) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/OwnerMetaMapper.java:84]: `TagMetadataObjectRelMapper` uses String instead of Enum type. 
Our PO uses String instead of Enum type. So we would better use String here.

File: core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java [hunk 28]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    if (!(store instanceof SupportsRelationOperations)) {
+      String errorMsg =
+          "OwnerManager cannot run with entity store that does not support relation operations, "
+              + "please configure the entity store to use relational entity store and restart the Gravitino server";
+      LOG.error(errorMsg);
+      throw new RuntimeException(errorMsg);
+    }
+    this.store = store;
```
[the remark this answers is not in the document]
Comment (qqqttt123) [reply core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java:64]: Accept the change.

File: core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+      throw new RuntimeException(errorMsg);
+    }
+  }
+
+  public void setOwner(
+      String metalake, MetadataObject metadataObject, String ownerName, Owner.Type ownerType) {
+    try {
+      NameIdentifier objectIdent = MetadataObjectUtil.toEntityIdent(metalake, metadataObject);
+      if (ownerType == Owner.Type.USER) {
+        NameIdentifier ownerIdent = AuthorizationUtils.ofUser(metalake, ownerName);
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java:71]: I would suggest using a different namespace for owner, not reuse "user".

Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/OwnerManager.java:71]: Owner isn't an entity. Owner is a relation. The relation will connect the two entities.

File: core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java [hunk 30]
```diff
@@ -360,6 +363,40 @@ public List<TagEntity> associateTagsWithMetadataObject(
...
+      NameIdentifier dstIdentifier,
+      Entity.EntityType dstType,
+      boolean override) {
+    if (relType == SupportsRelationOperations.Type.OWNER_REL) {
+      OwnerMetaService.getInstance().setOwner(srcIdentifier, srcType, dstIdentifier, dstType);
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Doesn't support the relation type %s", relType));
+    }
+  }
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java:398]: Change this to switch case.

Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java:400]: OK. I changed.

**Review by xunliu — APPROVED** [review]
LGTM
